packages/react-reconciler/src/ReactUpdateQueue.js

```diff
@@ -0,0 +1,639 @@
+  let workInProgressQueue = workInProgress.updateQueue;
+  if (workInProgressQueue === null) {
+    workInProgressQueue = workInProgress.updateQueue = createUpdateQueue(
+      workInProgress.memoizedState,
+    );
+  } else {
+    // TODO: I put this here rather than createWorkInProgress so that we don't
+    // clone the queue unnecessarily. There's probably a better way to
+    // structure this.
+    workInProgressQueue = ensureWorkInProgressQueueIsAClone(
+      workInProgress,
+      workInProgressQueue,
+    );
+  }
+
+  // Append the update to the end of the list.
+  if (workInProgressQueue.lastCapturedUpdate === null) {
+    // This is the first render phase update
+    workInProgressQueue.firstCapturedUpdate = workInProgressQueue.lastCapturedUpdate = update;
+  } else {
+    workInProgressQueue.lastCapturedUpdate.next = update;
+    workInProgressQueue.lastCapturedUpdate = update;
+  }
+  if (
+    workInProgressQueue.expirationTime === NoWork ||
+    workInProgressQueue.expirationTime > renderExpirationTime
+  ) {
+    // The incoming update has the earliest expiration of any update in the
+    // queue. Update the queue's expiration time.
+    workInProgressQueue.expirationTime = renderExpirationTime;
+  }
+}
+
+function ensureWorkInProgressQueueIsAClone<State>(
+  workInProgress: Fiber,
+  queue: UpdateQueue<State>,
+): UpdateQueue<State> {
+  const current = workInProgress.alternate;
+  if (current !== null) {
+    // If the work-in-progress queue is equal to the current queue,
+    // we need to clone it first.
+    if (queue === current.updateQueue) {
+      queue = workInProgress.updateQueue = cloneUpdateQueue(queue);
+    }
+  }
+  return queue;
+}
+
+function getStateFromUpdate<State>(
+  workInProgress: Fiber,
+  queue: UpdateQueue<State>,
+  update: Update<State>,
+  prevState: State,
+  nextProps: any,
+  instance: any,
+): any {
+  switch (update.tag) {
+    case ReplaceState: {
+      const payload = update.payload;
+      if (typeof payload === 'function') {
+        // Updater function
+        if (
+          debugRenderPhaseSideEffects ||
+          (debugRenderPhaseSideEffectsForStrictMode &&
+            workInProgress.mode & StrictMode)
+        ) {
+          payload.call(instance, prevState, nextProps);
+        }
+
+        return payload.call(instance, prevState, nextProps);
+      }
+      // State object
+      return payload;
+    }
+    case CaptureUpdate: {
+      workInProgress.effectTag =
+        (workInProgress.effectTag & ~ShouldCapture) | DidCapture;
+    }
+    // Intentional fallthrough
+    case UpdateState: {
+      const payload = update.payload;
+      let partialState;
+      if (typeof payload === 'function') {
+        // Updater function
+        if (
+          debugRenderPhaseSideEffects ||
+          (debugRenderPhaseSideEffectsForStrictMode &&
+            workInProgress.mode & StrictMode)
+        ) {
+          payload.call(instance, prevState, nextProps);
+        }
```

Comment:
This needs to be wrapped in DEV.

Reply:
We already have an `debugRenderPhaseSideEffects` check outside of a DEV block: https://github.com/facebook/react/blob/5dcf93d146ccde90af2c442d6cb32b29efddb46b/packages/react-reconciler/src/ReactFiberUpdateQueue.js#L297-L305

because `debugRenderPhaseSideEffects` is only true in DEV mode. But I agree we should put it in a DEV check anyway, for extra safety.

Reply:
Looks like our internal feature flag override isn't DEV aware, so it would happen in prod.